include/mfmg/amge.templates.hpp

```diff
@@ -302,6 +249,171 @@ void AMGe<dim, VectorType>::compute_restriction_sparse_matrix(
+unsigned int AMGe<dim, VectorType>::build_agglomerates_block(
+    std::array<unsigned int, dim> const &agglomerate_dim) const
+{
+  // Faces in deal.II are orderd as follows: left (x_m) = 0, right (x_p) = 1,
+  // front (y_m) = 2, back (y_p) = 3, bottom (z_m) = 4, top (z_p) = 5
+  unsigned int constexpr x_p = 1;
+  unsigned int constexpr y_p = 3;
+  unsigned int constexpr z_p = 5;
+
+  // Flag the cells to create the agglomerates
+  unsigned int agglomerate = 1;
+  for (auto cell : _dof_handler.active_cell_iterators())
+  {
+    if ((cell->is_locally_owned()) && (cell->user_index() == 0))
+    {
+#if MFMG_DEBUG
+      int const cell_level = cell->level();
+#endif
+      cell->set_user_index(agglomerate);
+      auto current_z_cell = cell;
+      unsigned int const d_3 = (dim < 3) ? 1 : agglomerate_dim.back();
+      for (unsigned int k = 0; k < d_3; ++k)
+      {
+        auto current_y_cell = current_z_cell;
+        for (unsigned int j = 0; j < agglomerate_dim[1]; ++j)
+        {
+          auto current_cell = current_y_cell;
+          for (unsigned int i = 0; i < agglomerate_dim[0]; ++i)
+          {
+            current_cell->set_user_index(agglomerate);
+            if (current_cell->at_boundary(x_p) == false)
+            {
+              // TODO For now, we assume that there is no adaptive refinement.
+              // When we change this, we will need to switch to hp::DoFHandler
+              auto neighbor_cell = current_cell->neighbor(x_p);
+#if MFMG_DEBUG
+              if ((!neighbor_cell->active()) ||
+                  (neighbor_cell->level() != cell_level))
+                throw std::runtime_error("Mesh locally refined");
+#endif
+              if (neighbor_cell->is_locally_owned())
+                current_cell = neighbor_cell;
+            }
+            else
+              break;
+          }
+          if (current_y_cell->at_boundary(y_p) == false)
+          {
+            auto neighbor_y_cell = current_y_cell->neighbor(y_p);
+            if (neighbor_y_cell->is_locally_owned())
+            {
+#if MFMG_DEBUG
+              if ((!neighbor_y_cell->active()) ||
+                  (neighbor_y_cell->level() != cell_level))
+                throw std::runtime_error("Mesh locally refined");
+#endif
+              current_y_cell = neighbor_y_cell;
+            }
+          }
+          else
+            break;
+        }
+        if ((dim == 3) && (current_z_cell->at_boundary(z_p) == false))
+        {
+          auto neighbor_z_cell = current_z_cell->neighbor(z_p);
+          if (neighbor_z_cell->is_locally_owned())
+          {
+#if MFMG_DEBUG
+            if ((!neighbor_z_cell->active()) ||
+                (neighbor_z_cell->level() != cell_level))
+              throw std::runtime_error("Mesh locally refined");
+#endif
+            current_z_cell = neighbor_z_cell;
+          }
+        }
+        else
+          break;
+      }
+
+      ++agglomerate;
+    }
+  }
+
+  return agglomerate - 1;
```

Comment:
Is that because you meant to sent non-zero user indices?

Reply:
Yes, I just wanted an easy way to check that the cells were flagged correctly that's why I didn't want to use zero which is the default value of the user index